compiler/qsc_parse/src/lex/cooked.rs

```diff
@@ -357,15 +370,26 @@ impl<'a> Lexer<'a> {
     }
 
     #[allow(clippy::too_many_lines)]
-    fn single(&mut self, single: Single) -> Result<TokenKind, Error> {
+    fn single(&mut self, single: Single, span: Span) -> Result<TokenKind, Error> {
         match single {
             Single::Amp => {
                 let op = ClosedBinOp::AmpAmpAmp;
                 self.expect(Single::Amp, TokenKind::ClosedBinOp(op))?;
                 self.expect(Single::Amp, TokenKind::ClosedBinOp(op))?;
                 Ok(self.closed_bin_op(op))
             }
-            Single::Apos => Ok(TokenKind::Apos),
+            Single::Apos => {
+                // disallow whitespace and comments after apostrophes
+                // see https://github.com/microsoft/qsharp/issues/407 for context
+                if matches!(
+                    self.tokens.peek().map(|tok| tok.kind),
+                    Some(raw::TokenKind::Whitespace | raw::TokenKind::Comment(_))
+                ) {
+                    Err(Error::UnfinishedGeneric(span))
+                } else {
+                    Ok(TokenKind::Apos)
+                }
```

Comment:
I wouldn't think of this in terms of the negative cases (it's an error for an Apos to be followed by a Whitespace or Comment). There are lots and lots of negative cases, but there's only one positive case: an Apos token must be followed by an Ident. Let's cook both of them into a single token representing a type parameter name. ([Parse, don't validate.](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/))

This does mean that the name of all type parameters will change downstream to include the `'`, but I think that's fine.